required/latex-lab/socket-sample.tex

```diff
@@ -0,0 +1,411 @@
+%
+In a nutshell: instead of having a fixed code block somewhere as part
+of the code that implements a certain programming logic, there is a named socket
+which is assigned a chunk of code to be executed in this place.
+
+This is done by first declaring the socket with:
+\begin{quote}
+\cs{NewSocket}\marg{socket-name}\marg{number-of-arguments}
+\end{quote}
+This is then used in code through
+\begin{quote}
+ \cs{UseSocket}\marg{socket-name}
+\end{quote}
+or, if the socket takes a number of arguments with
+\begin{quote}
+  \cs{UseSocket}\marg{socket-name}\marg{arg\textsubscript{1}}\ldots
+  \marg{arg\textsubscript{number-of-arguments}}
+\end{quote}
+
+In addition, one or several named chunks of code implementing different logic for this
+socket are set up, each with a declaration of the form:
+\begin{quote}
+  \cs{DeclareSocketCode}\marg{socket-name}\marg{chunk-name}\marg{code})
+\end{quote}
+Finally,
+one of them is assigned to the socket:
+\begin{quote}
+\cs{AssignSocketCode}\marg{socket-name}\marg{chunk-name}
+\end{quote}
+If the programming logic should change, then all that is necessary is
+to make a new assignment with \cs{AssignSocketCode} to a different
+\marg{chunk-name}.
+
+If the socket takes arguments, then those need to be provided to
+\cs{UseSocket} and in that case they can be referenced in the \meta{code}
+argument of \cs{DeclareSocketCode} with \verb=#1=, \verb=#2=, etc.
+
+In most cases a named socket is used only in a single place, but there
+is, of course, nothing wrong wth using it in several places, as long
+as the code in all places is supposed to change in the same way if the socket gets reassigned a different chunk of code.
+
+
+\section{Alternative names}
+
+
+By the way:
+\begin{itemize}
+\item
+  should it be \texttt{gassign} ? or some other verb?
+\item
+  And what about \cs{socket\_use:n} ? Should that be \texttt{:nw}
+  as wedon't really control the arguments or should we have
+  \cs{socket\_use:nn}, \cs{socket\_use:nnn}, \ldots, just to make it
+  clearer how many arguments are (supposed to be) picked up?
+\end{itemize}
+
+
+
+
+\section{Example and testing}
+
+From here on we scroll through errors because the example expliitly generate a few:
+\begin{verbatim}
+  \scrollmode
+\end{verbatim}
+\scrollmode
+
+We declare a new socket named \texttt{foo} expecting 2 arguments:
+\begin{verbatim}
+  \NewSocket{foo}{2}
+\end{verbatim}
+\NewSocket{foo}{2}
+
+Such a declaration has do be unique across the whole \LaTeX{} run so
+if another package attempts to use the same name (regardless of the
+number of arguments) it will generate an error:
+\begin{verbatim}
+  \NewSocket{foo}{2}    % Error (already declared)
+  \NewSocket{foo}{1}    % Error (already declared)
+\end{verbatim}
+\NewSocket{foo}{2}
+\NewSocket{foo}{1}
+
+You also get an error if you attempt to declare some socket code and the
+socket name is not yet defined, e.g.,
+\begin{verbatim}
+  \DeclareSocketCode{baz}{undeclared}{} % Error (Socket not declared)
```

Comment:
Should this error, or just store the code?  For example if, say, `multicols` defines a socket `outputcolumn`, and another package `fancycols`, which is loaded before, adds some code to that socket?  Of course this could be solved by doing `\DeclareSocketCode` in the hook `package/multicols/after`, but it's kinda workaroundish, I think.

Reply:
Good point, probably it should show the same behavior even though the number of use cases will be far less compared to hooks, but I guess they aren't zero. So perhaps I should change that.

Reply:
Thinking about this a bit more I'm not so sure that it is a good idea after all. You would need to do the same then to `\AssignSocketPlug` and then the  question is what happens you have  two such assignments before the socket is declared? Who is then going to win? The one assigned to the socket when it is declared "noop" ie the last assignment that was made? That would be logical but not helpful. And how would you adjust the situation?

If you go via `package/multicol/after` then there is a clear intend  that you overwrite and if two packages do this you could order the decision with a rule if necessary. So for now I'm inclined to leave it as it is and see what use case appear in reality.